From 880f77b32788ecdc2d8907749878b12231015587 Mon Sep 17 00:00:00 2001
From: Hani Kazmi <hanikazmi@google.com>
Date: Tue, 27 Sep 2022 10:19:45 +0000
Subject: [PATCH] Update Parcel readLazyValue to ignore negative object lengths

Addresses a security vulnerability where a (-8) length object would
cause dataPosition to be reset back to the statt of the value, and be
re-read again.

Bug: 240138294
Test: atest ParcelTest BundleTest AmbiguousBundlesTest
Test: manually ran PoC
Change-Id: I1ab1df6f2a802d8cdf02c89c12959b09d7b1a5c4
Merged-In: I1ab1df6f2a802d8cdf02c89c12959b09d7b1a5c4
(cherry picked from commit 34683275498914ece5ee9435846b7b429ccfc964)
Merged-In: I1ab1df6f2a802d8cdf02c89c12959b09d7b1a5c4
---
 core/java/android/os/Parcel.java | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/core/java/android/os/Parcel.java b/core/java/android/os/Parcel.java
index 2664f05a9664..85bf6f6f0bc8 100644
--- a/core/java/android/os/Parcel.java
+++ b/core/java/android/os/Parcel.java
@@ -4353,6 +4353,9 @@ public final class Parcel {
         int type = readInt();
         if (isLengthPrefixed(type)) {
             int objectLength = readInt();
+            if (objectLength < 0) {
+                return null;
+            }
             int end = MathUtils.addOrThrow(dataPosition(), objectLength);
             int valueLength = end - start;
             setDataPosition(end);
-- 
2.38.1.273.g43a17bfeac-goog

